File: 2020/199.py
"""
see https://leetcode.com/problems/binary-tree-right-side-view/
"""
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        # similar to bfs
        if not root: return []

        res = []
        stack = [root]

        while stack:
            nextLayer = []
            for nd in stack:
                if nd.left:
                    nextLayer += nd.left,
                if nd.right:
                    nextLayer += nd.right,
            res += stack[-1].val,
            stack = nextLayer

        return res

File: current_session/c++/199.cpp
 
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

//  Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        // check layer by layer
        // use stack
        // queue<TreeNode*>mQ;
        vector<TreeNode*> layer;
        layer.push_back(root);
        vector<int> res;
        while (layer.size() > 0) {
            TreeNode nd = *layer.at(layer.size()-1);
            res.push_back(nd.val);
            vector<TreeNode*> nextLayer;
            for (int i = 0; i < layer.size(); i++) {
                TreeNode x = *layer.at(i);
                if (x.left) nextLayer.push_back(x.left);
                if (x.right) nextLayer.push_back(x.right);
            }
            layer = nextLayer;
        }
        return res;
    }
};
